src/pm-conversations/pmConversationsSelectors.js

```diff
@@ -51,10 +67,114 @@ export const getRecentConversations: Selector<PmConversationData[]> = createSele
+      ids: pmUnreadsKeyFromMessage(msg, ownUser.user_id),
+      recipients: normalizeRecipientsSansMe(msg.display_recipient, ownUser.email),
+      msgId: msg.id,
     }));
+
+    const sortedByMostRecent = collateByRecipient(items);
+
+    return attachUnread(sortedByMostRecent);
   },
 );
 
+/**
+ * Modern implementation of {@link getRecentConversations}. Returns exactly the
+ * most recent conversations. Requires server-side support.
+ */
+const getRecentConversationsImpl: Selector<PmConversationData[]> = createSelector(
+  getOwnUser,
+  getAllUsersById,
+  getRecentPrivateConversations,
+  getAttachUnread,
+  (
+    ownUser: User,
+    allUsers: Map<number, UserOrBot>,
+    recentPCs: RecentPrivateConversation[],
+    attachUnread: UnreadAttacher,
+  ) => {
+    const getEmail = (id: number): string => {
+      const user = allUsers.get(id);
+      if (user) {
+        return user.email;
+      }
+      throw new Error('getRecentConversations: unknown user id');
+    };
+
+    const recipients = recentPCs.map(conversation => {
+      const userIds = conversation.user_ids.slice();
+      if (userIds.length !== 1) {
+        userIds.push(ownUser.user_id);
+        userIds.sort((a, b) => a - b);
+      }
+
+      return {
+        ids: userIds.join(','),
+        recipients: normalizeRecipientsSansMe(
+          userIds.map(id => ({ email: getEmail(id) })),
+          ownUser.email,
+        ),
+        msgId: conversation.max_message_id,
+      };
+    });
+
+    return attachUnread(recipients);
+  },
+);
+
+/**
+ * The server version in which 'recent_private_conversations' was first made
+ * available.
+ */
+const DIVIDING_LINE = new ZulipVersion('2.1-dev-384-g4c3c669b41');
+
+// Private. Selector to choose between other selectors. (This avoids needlessly
+// recomputing the old version when we're on a new server, or vice versa.)
+const getMetaselector: Selector<Selector<PmConversationData[]>> = createSelector(
+  getServerVersion,
+  version => {
+    // If we're talking to a new enough version of the Zulip server, we don't
+    // need the legacy impl; the modern one will always return a superset of
+    // its content.
+    if (version && version.isAtLeast(DIVIDING_LINE)) {
+      return getRecentConversationsImpl;
+    }
+
+    // If we're _not_ talking to a newer version of the Zulip server, then
+    // there's no point in using the modern version; it will only return
+    // messages received in the current session, which should all be in the
+    // legacy impl's data as well.
+    return getRecentConversationsLegacyImpl;
+  },
+);
+
+/**
+ * Get a list of the most recent private conversations, including the most
+ * recent message from each.
+ *
+ * Switches between implementations as appropriate for the current server
+ * version.
+ */
+export const getRecentConversations: Selector<PmConversationData[]> = createSelector(
+  state => state,
+  getMetaselector,
+  (state, metaselector) => metaselector(state),
+);
```

Comment:
This is an example of a selector that doesn't do any usefully cacheable work, so there's no need for `createSelector`.

(Also the `state => state` means the cache will never be hit unless `state` is identically unchanged, and in that case we generally shouldn't be getting this far anyway.)